src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -356,5 +418,220 @@ private void RequestExecutionQueue_Errored(object sender, RequestShutdownEventAr
+                // 2.  There are diagnostics from the current document.  Store the diagnostics for the fileUri and document
+                //      so they can be published along with contributions to the fileUri from other documents.
+                // 3.  There are no diagnostics contributed by this document to the fileUri (could be some from other documents).
+                //     We should clear out the diagnostics for this document for the fileUri.
+                if (diagnostics.IsEmpty)
+                {
+                    // We published an empty set of diagnostics for this uri.  We no longer need to keep track of this mapping
+                    // since there will be no previous diagnostics that we need to clear out.
+                    _documentsToPublishedUris.MultiRemove(document.Id, fileUri);
+
+                    // There are not any diagnostics to keep track of for this file, so we can stop.
+                    _publishedFileToDiagnostics.Remove(fileUri);
+                }
+                else if (fileUriToDiagnostics.ContainsKey(fileUri))
+                {
+                    // We do have diagnostics from the current document - update the published diagnostics map
+                    // to contain the new diagnostics contributed by this document for this uri.
+                    var documentsToPublishedDiagnostics = _publishedFileToDiagnostics.GetOrAdd(fileUri, (_) =>
+                        new Dictionary<DocumentId, ImmutableArray<LSP.Diagnostic>>());
+                    documentsToPublishedDiagnostics[document.Id] = fileUriToDiagnostics[fileUri];
+                }
+                else
+                {
+                    // There were diagnostics from other documents, but none from the current document.
+                    // If we're tracking the current document, we can stop.
+                    _publishedFileToDiagnostics.GetOrDefault(fileUri)?.Remove(document.Id);
+                    _documentsToPublishedUris.MultiRemove(document.Id, fileUri);
+                }
+            }
+        }
+
+        private async Task SendDiagnosticsNotificationAsync(Uri uri, ImmutableArray<LSP.Diagnostic> diagnostics)
+        {
+            var publishDiagnosticsParams = new PublishDiagnosticParams { Diagnostics = diagnostics.ToArray(), Uri = uri };
+            await _jsonRpc.NotifyWithParameterObjectAsync(Methods.TextDocumentPublishDiagnosticsName, publishDiagnosticsParams).ConfigureAwait(false);
+        }
+
+        private async Task<Dictionary<Uri, ImmutableArray<LSP.Diagnostic>>> GetDiagnosticsAsync(Document document, CancellationToken cancellationToken)
+        {
+            var diagnostics = _diagnosticService.GetDiagnostics(document.Project.Solution.Workspace, document.Project.Id, document.Id, id: null, includeSuppressedDiagnostics: false, forPullDiagnostics: false, cancellationToken)
```

Comment:
note that this says `forPushDiagnostics: true`. meaning that we will push in our *current* mode where we have not switched to 'pull diagnostics'.  However, if we flip the new option to "use pull diagnostics", this will now cause us to get no diagnostics back here, so we'll push nothing.